packages/next/client/link.tsx

```diff
@@ -184,28 +179,29 @@ function Link(props: React.PropsWithChildren<LinkProps>) {
 
   const [childElm, setChildElm] = React.useState<Element>()
 
-  const { href, as } = React.useMemo(
-    () => ({
-      href: formatUrl(props.href),
-      as: props.as ? formatUrl(props.as) : props.as,
-    }),
-    [props.href, props.as]
-  )
+  const router = useRouter()
+
+  const { href, as } = React.useMemo(() => {
+    const resolvedHref = resolve(router.pathname, formatUrl(props.href))
+    return {
+      href: resolvedHref,
+      as: props.as
+        ? resolve(router.pathname, formatUrl(props.as))
+        : resolvedHref,
+    }
+  }, [router.pathname, props.href, props.as])
```

Comment:
This line breaks the component if rendered outside `<RouteContext>`, since `router` is `undefined`.

See #15543